src/util/address.rs

```diff
@@ -406,6 +406,150 @@ impl Payload {
+    }
+
+    /// Create a witness pay to public key payload from a public key
+    pub fn p2wpkh(pk: &ecdsa::PublicKey) -> Result<Payload, Error> {
+        if !pk.compressed {
+            return Err(Error::UncompressedPubkey);
+        }
+
+        let mut hash_engine = WPubkeyHash::engine();
+        pk.write_into(&mut hash_engine)
+            .expect("engines don't error");
+
+        Ok(Payload::WitnessProgram {
+            version: WitnessVersion::V0,
+            program: WPubkeyHash::from_engine(hash_engine)[..].to_vec(),
+        })
+    }
+
+    /// Create a pay to script payload that embeds a witness pay to public key
+    pub fn p2shwpkh(pk: &ecdsa::PublicKey) -> Result<Payload, Error> {
+        if !pk.compressed {
+            return Err(Error::UncompressedPubkey);
+        }
+
+        let mut hash_engine = WPubkeyHash::engine();
+        pk.write_into(&mut hash_engine)
+            .expect("engines don't error");
+
+        let builder = script::Builder::new()
+            .push_int(0)
+            .push_slice(&WPubkeyHash::from_engine(hash_engine)[..]);
+
+        Ok(Payload::ScriptHash(ScriptHash::hash(
+            builder.into_script().as_bytes(),
+        )))
+    }
+
+    /// Create a witness pay to script hash payload.
+    pub fn p2wsh(script: &script::Script) -> Payload {
+        Payload::WitnessProgram {
+            version: WitnessVersion::V0,
+            program: WScriptHash::hash(&script[..])[..].to_vec(),
+        }
+    }
+
+    /// Create a pay to script payload that embeds a witness pay to script hash address
+    pub fn p2shwsh(script: &script::Script) -> Payload {
+        let ws = script::Builder::new()
+            .push_int(0)
+            .push_slice(&WScriptHash::hash(&script[..])[..])
+            .into_script();
+
+        Payload::ScriptHash(ScriptHash::hash(&ws[..]))
+    }
+
+    /// Create a pay to taproot payload from untweaked key
+    pub fn p2tr<C: Verification>(
+        secp: Secp256k1<C>,
+        internal_key: UntweakedPublicKey,
+        merkle_root: Option<TapBranchHash>,
+    ) -> Payload {
+        Payload::WitnessProgram {
+            version: WitnessVersion::V1,
+            program: internal_key
+                .tap_tweak(secp, merkle_root)
+                .into_inner()
+                .serialize()
+                .to_vec(),
+        }
+    }
+
+    /// Create a pay to taproot payload from a pre-tweaked output key.
+    ///
+    /// This method is not recommended for use and [Payload::p2tr()] should be used where possible.
+    pub fn p2tr_tweaked(output_key: TweakedPublicKey) -> Payload {
+        Payload::WitnessProgram {
+            version: WitnessVersion::V1,
+            program: output_key.into_inner().serialize().to_vec(),
+        }
+    }
+}
+
+/// A utility struct to encode an address payload with the given parameters.
+pub struct AddressEncoding<'a, 'b> {
+    /// The address payload to encode.
+    pub payload: &'a Payload,
```

Comment:
Better to have all fields private as well

Reply:
Same, it is supposed to be public so library users can encode a Payload using their own params.

Reply:
Hm, then I can do `let encoding = AddressEncoding { p2pkh_prefix: 0x88, p2sh_prefix: 0x88, bech32_hrp: "crazy" };` which will produce an invalid address strings. Why user would need such API?

Reply:
Can provide e.g. Litecoin params. The linked issue explains the context: https://github.com/rust-bitcoin/rust-bitcoin/issues/642

Reply:
In terms of API it is always better to trait-extend existing type with more constructors than to allow end-users to arbitrary change field values at any time. I.e. I still think it will be nicer to have

```rust
pub struct AddressEncoding<'a, 'b> {
   payload: &'a Payload,
   pkh_prefix: u8,
   sh_prefix: u8,
   bech32_hrp: &'b str,
}

impl<'a, 'b> AddressEncoding<'a, 'b> {
   pub from_network(network: Network) -> AddressEncoding<'a, 'b> {
      // ...
   }
}
```

and in downstream "Litecoin" library

```rust
pub trait FromLitecoinNetwork<'a, 'b> {
  fn from_network(network: litecoin::Network) -> AddressEncoding<'a, 'b>;
}

impl<'a, 'b> FromLitecoinNetwork<'a, 'b> for AddressEncoding<'a, 'b> {
  // ...
} 
```

Reply:
I don't quite understand, why is `FromLitecoinNetwork` a trait, and how does the impl actually construct the `AddressEncoding` struct if the fields are private? Seems like another public constructor is needed that takes all the 4 struct members, right? At this point it seems like a whole bunch of boilerplate that achieves the same thing and gives no more safety to the user. What am I missing?

Reply:
You are right that you would not be able to access private fields from trait impl in other crate; but the idea is that fields must not be changed individually _after_ the struct creation. Thus, having public constructor filling all private fields at struct init, used by more specific constructors added via traits in descendant crates is a better API, preventing arbitrary modification of the fields in the instantiated struct.

Reply:
I don't see why changing would be a problem. A garbage value can be created in constructor too.

However I do see one reason to have the struct `non_exhaustive` (emulated by a private field if needed): possibility of adding new fields. (Presumably with new address formats in the future.) Not sure how much it is relevant though.

Reply:
@dr-orlovsky I like protecting library users, but I don't think your proposal does that. If there is a public constructor, a user can always change an individual field by creating a new instance changing the field they like.

I propose to keep it as is, as the added boilerplate does not really protect library users but increases complexity.

I added this comment to the struct doc to warn library users instead:

 `/// This is a low-level utility struct. Consider using `Address` instead.`

Reply:
@Kixunil I believe `non_exhaustive` would disallow constructing the struct outside of the crate directly, so a constructor would be needed. But then, the constructor would have to be extended instead. I don't think it's relevant anyway.

Reply:
The constructor could just assign a default that could be overwritten separately. Agree it's most likely not useful.

That doc comment is a good idea!

Reply:
Ok, let's leave it as it is now